File: facecat_java/src/facecat/topin/div/FCToolTip.java
/*捂脸猫FaceCat框架 v1.0
 1.创始人-矿洞程序员-上海宁米科技创始人-脉脉KOL-陶德 (微信号:suade1984);
 2.联合创始人-上海宁米科技创始人-袁立涛(微信号:wx627378127);
 3.联合创始人-河北思尔企业管理咨询有限公司合伙人-肖添龙(微信号:xiaotianlong_luu);
 4.联合开发者-陈晓阳(微信号:chenxiaoyangzxy)，助理-朱炜(微信号:cnnic_zhu);
 5.该框架开源协议为BSD，欢迎对我们的创业活动进行各种支持，欢迎更多开发者加入。
 包含C/C++,Java,C#,iOS,MacOS,Linux六个版本的图形和通讯服务框架。
 */
package facecat.topin.div;

import facecat.topin.core.*;
import java.util.*;

/**
 * 提示标签
 */
public class FCToolTip extends FCView {

    /**
     * 创建标签
     */
    public FCToolTip() {
        setAutoSize(true);
        setBackColor(FCColor.argb(255, 255, 40));
        setBorderColor(FCColor.Border);
        setTopMost(true);
        setVisible(false);
    }

    /**
     * 上一次触摸的位置
     */
    private FCPoint m_lastTouchPoint;

    /**
     * 秒表ID
     */
    private int m_timerID = getNewTimerID();

    /**
     * 剩余保留显示毫秒数
     */
    protected int m_remainAutoPopDelay;

    /**
     * 剩余延迟显示毫秒数
     */
    protected int m_remainInitialDelay;

    protected int m_autoPopDelay = 5000;

    /**
     * 提示显示的时间长度
     */
    public int autoPopDelay() {
        return m_autoPopDelay;
    }

    /**
     * 保留显示的时间长度
     */
    public void setAutoPopDelay(int autoPopDelay) {
        m_autoPopDelay = autoPopDelay;
    }

    protected int m_initialDelay = 500;

    /**
     * 获取触摸静止时延迟显示的毫秒数
     */
    public int initialDelay() {
        return m_initialDelay;
    }

    /**
     * 设置触摸静止时延迟显示的毫秒数
     */
    public void setinitialDelay(int initialDelay) {
        m_initialDelay = initialDelay;
    }

    protected boolean m_showAlways;

    /**
     * 获取是否总是显示
     */
    public boolean showAlways() {
        return m_showAlways;
    }

    /**
     * 设置是否总是显示
     */
    public void setshowAlways(boolean showAlways) {
        m_showAlways = showAlways;
    }

    protected boolean m_useAnimation;

    /**
     * 获取是否在显示隐藏时使用动画
     */
    public boolean useAnimation() {
        return m_useAnimation;
    }

    /**
     * 设置是否在显示隐藏时使用动画
     */
    public void setuseAnimation(boolean useAnimation) {
        m_useAnimation = useAnimation;
    }

    /**
     * 销毁控件方法
     */
    @Override
    public void delete() {
        if (!isDeleted()) {
            stopTimer(m_timerID);
        }
        super.delete();
    }

    /**
     * 获取控件类型
     */
    @Override
    public String getControlType() {
        return "ToolTip";
    }

    /**
     * 获取属性值
     *
     * @param name 属性名称
     * @param value 返回属性值
     * @param type 返回属性类型
     */
    @Override
    public void getProperty(String name, RefObject<String> value, RefObject<String> type) {
        if (name.equals("autopopupdelay")) {
            type.argvalue = "int";
            value.argvalue = FCStr.convertIntToStr(autoPopDelay());
        } else if (name.equals("initialdelay")) {
            type.argvalue = "int";
            value.argvalue = FCStr.convertIntToStr(initialDelay());
        } else if (name.equals("showalways")) {
            type.argvalue = "bool";
            value.argvalue = FCStr.convertBoolToStr(showAlways());
        } else if (name.equals("useanimation")) {
            type.argvalue = "bool";
            value.argvalue = FCStr.convertBoolToStr(useAnimation());
        } else {
            super.getProperty(name, value, type);
        }
    }

    /**
     * 获取属性名称列表
     */
    public ArrayList<String> getPropertyNames() {
        ArrayList<String> propertyNames = super.getPropertyNames();
        propertyNames.addAll(Arrays.asList(new String[]{"AutoPopupDelay", "InitialDelay", "ShowAlways", "UseAnimation"}));
        return propertyNames;
    }


    /**
     * 隐藏控件
     */
    @Override
    public void hide() {
        setVisible(false);
    }

    /**
     * 添加控件方法
     */
    @Override
    public void onLoad() {
        super.onLoad();
        m_lastTouchPoint = getTouchPoint();
        startTimer(m_timerID, 10);
    }

    /**
     * 秒表方法
     *
     * @param timerID 秒表ID
     */
    @Override
    public void onTimer(int timerID) {
        super.onTimer(timerID);
        if (m_timerID == timerID) {
            FCPoint mp = getTouchPoint();
            if (!m_showAlways) {
                if (m_lastTouchPoint.x != mp.x || m_lastTouchPoint.y != mp.y) {
                    setVisible(false);
                }
            }
            m_lastTouchPoint = mp;
            if (m_remainAutoPopDelay > 0) {
                m_remainAutoPopDelay -= 10;
                if (m_remainAutoPopDelay <= 0) {
                    setVisible(false);
                }
            }
            if (m_remainInitialDelay > 0) {
                m_remainInitialDelay -= 10;
                if (m_remainInitialDelay <= 0) {
                    setVisible(true);
                }
            }
        }
    }

    /**
     * 可见状态改变方法
     */
    @Override
    public void onVisibleChanged() {
        super.onVisibleChanged();
        if (m_native != null) {
            if (isVisible()) {
                m_native.addControl(this);
                m_remainAutoPopDelay = m_autoPopDelay;
                m_remainInitialDelay = 0;
            } else {
                m_native.removeControl(this);
                startTimer(m_timerID, 10);
                m_remainAutoPopDelay = 0;
                m_remainInitialDelay = 0;
            }
            getNative().invalidate();
        }
    }

    /**
     * 设置属性
     *
     * @param name 属性名称
     * @param value 属性值
     */
    @Override
    public void setProperty(String name, String value) {
        if (name.equals("autopopupdelay")) {
            setAutoPopDelay(FCStr.convertStrToInt(value));
        } else if (name.equals("initialdelay")) {
            setinitialDelay(FCStr.convertStrToInt(value));
        } else if (name.equals("showalways")) {
            setshowAlways(FCStr.convertStrToBool(value));
        } else if (name.equals("useanimation")) {
            setuseAnimation(FCStr.convertStrToBool(value));
        } else {
            super.setProperty(name, value);
        }
    }

    /**
     * 显示控件
     */
    public void show() {
        m_remainAutoPopDelay = 0;
        m_remainInitialDelay = m_initialDelay;
        setVisible(m_initialDelay == 0);
        getNative().invalidate();
    }
}

File: facecat_cs/div/FCToolTip.cs
/*捂脸猫FaceCat框架 v1.0
 1.创始人-矿洞程序员-上海宁米科技创始人-脉脉KOL-陶德 (微信号:suade1984);
 2.联合创始人-上海宁米科技创始人-袁立涛(微信号:wx627378127);
 3.联合创始人-河北思尔企业管理咨询有限公司合伙人-肖添龙(微信号:xiaotianlong_luu);
 4.联合开发者-陈晓阳(微信号:chenxiaoyangzxy)，助理-朱炜(微信号:cnnic_zhu);
 5.该框架开源协议为BSD，欢迎对我们的创业活动进行各种支持，欢迎更多开发者加入。
 包含C/C++,Java,C#,iOS,MacOS,Linux六个版本的图形和通讯服务框架。
 */

using System;
using System.Collections.Generic;

namespace FaceCat {
    /// <summary>
    /// 提示标签
    /// </summary>
    public class FCToolTip : FCLabel {
        /// <summary>
        /// 创建标签
        /// </summary>
        public FCToolTip() {
            AutoSize = true;
            BackColor = FCColor.argb(255, 255, 40);
            BorderColor = FCColor.Border;
            TopMost = true;
            Visible = false;
        }

        /// <summary>
        /// 上一次触摸的位置
        /// </summary>
        private FCPoint m_lastTouchPoint;

        /// <summary>
        /// 秒表ID
        /// </summary>
        private int m_timerID = getNewTimerID();

        /// <summary>
        /// 剩余保留显示毫秒数
        /// </summary>
        protected int m_remainAutoPopDelay;

        /// <summary>
        /// 剩余延迟显示毫秒数
        /// </summary>
        protected int m_remainInitialDelay;

        protected int m_autoPopDelay = 5000;

        /// <summary>
        /// 提示保留显示的时间长度
        /// </summary>
        public virtual int AutoPopDelay {
            get { return m_autoPopDelay; }
            set { m_autoPopDelay = value; }
        }

        protected int m_initialDelay = 500;

        /// <summary>
        /// 获取或设置触摸静止时延迟显示的毫秒数
        /// </summary>
        public virtual int InitialDelay {
            get { return m_initialDelay; }
            set { m_initialDelay = value; }
        }

        protected bool m_showAlways;

        /// <summary>
        /// 获取或设置是否总是显示
        /// </summary>
        public virtual bool ShowAlways {
            get { return m_showAlways; }
            set { m_showAlways = value; }
        }

        protected bool m_useAnimation;

        /// <summary>
        /// 获取或设置是否在显示隐藏时使用动画
        /// </summary>
        public virtual bool UseAnimation {
            get { return m_useAnimation; }
            set { m_useAnimation = value; }
        }

        /// <summary>
        /// 销毁控件方法
        /// </summary>
        public override void delete() {
            if (!IsDeleted) {
                stopTimer(m_timerID);
            }
            base.delete();
        }

        /// <summary>
        /// 获取控件类型
        /// </summary>
        /// <returns></returns>
        public override String getControlType() {
            return "ToolTip";
        }

        /// <summary>
        /// 获取属性值
        /// </summary>
        /// <param name="name">属性名称</param>
        /// <param name="value">返回属性值</param>
        /// <param name="type">返回属性类型</param>
        public override void getProperty(String name, ref String value, ref String type) {
            if (name == "autopopupdelay") {
                type = "int";
                value = FCStr.convertIntToStr(AutoPopDelay);
            }
            else if (name == "initialdelay") {
                type = "int";
                value = FCStr.convertIntToStr(InitialDelay);
            }
            else if (name == "showalways") {
                type = "bool";
                value = FCStr.convertBoolToStr(ShowAlways);
            }
            else if (name == "useanimation") {
                type = "bool";
                value = FCStr.convertBoolToStr(UseAnimation);
            }
            else {
                base.getProperty(name, ref value, ref type);
            }
        }

        /// <summary>
        /// 获取属性名称列表
        /// </summary>
        /// <returns>属性名称列表</returns>
        public override ArrayList<String> getPropertyNames() {
            ArrayList<String> propertyNames = base.getPropertyNames();
            propertyNames.AddRange(new String[] { "AutoPopupDelay", "InitialDelay", "ShowAlways", "UseAnimation" });
            return propertyNames;
        }

        /// <summary>
        /// 隐藏控件
        /// </summary>
        public override void hide() {
            Visible = false;
        }

        /// <summary>
        /// 添加控件方法
        /// </summary>
        public override void onLoad() {
            base.onLoad();
            m_lastTouchPoint = TouchPoint;
            startTimer(m_timerID, 10);
        }

        /// <summary>
        /// 秒表方法
        /// </summary>
        /// <param name="timerID">秒表ID</param>
        public override void onTimer(int timerID) {
            base.onTimer(timerID);
            if (m_timerID == timerID) {
                FCPoint mp = TouchPoint;
                if (!m_showAlways) {
                    if (m_lastTouchPoint.x != mp.x || m_lastTouchPoint.y != mp.y) {
                        Visible = false;
                    }
                }
                m_lastTouchPoint = mp;
                if (m_remainAutoPopDelay > 0) {
                    m_remainAutoPopDelay -= 10;
                    if (m_remainAutoPopDelay <= 0) {
                        Visible = false;
                    }
                }
                if (m_remainInitialDelay > 0) {
                    m_remainInitialDelay -= 10;
                    if (m_remainInitialDelay <= 0) {
                        Visible = true;
                    }
                }
            }
        }

        /// <summary>
        /// 可见状态改变方法
        /// </summary>
        public override void onVisibleChanged() {
            base.onVisibleChanged();
            if (m_native != null) {
                if (Visible) {
                    m_native.addControl(this);
                    m_remainAutoPopDelay = m_autoPopDelay;
                    m_remainInitialDelay = 0;
                }
                else {
                    m_native.removeControl(this);
                    startTimer(m_timerID, 10);
                    m_remainAutoPopDelay = 0;
                    m_remainInitialDelay = 0;
                }
                Native.invalidate();
            }
        }

        /// <summary>
        /// 设置属性
        /// </summary>
        /// <param name="name">属性名</param>
        /// <param name="value">属性值</param>
        public override void setProperty(String name, String value) {
            if (name == "autopopupdelay") {
                AutoPopDelay = FCStr.convertStrToInt(value);
            }
            else if (name == "initialdelay") {
                InitialDelay = FCStr.convertStrToInt(value);
            }
            else if (name == "showalways") {
                ShowAlways = FCStr.convertStrToBool(value);
            }
            else if (name == "useanimation") {
                UseAnimation = FCStr.convertStrToBool(value);
            }
            else {
                base.setProperty(name, value);
            }
        }

        /// <summary>
        /// 显示控件
        /// </summary>
        public override void show() {
            m_remainAutoPopDelay = 0;
            m_remainInitialDelay = m_initialDelay;
            Visible = m_initialDelay == 0;
            Native.invalidate();
        }
    }
}
